sql: add cluster setting for maximum read buffer size

otan opened the pull request:
Resolves #50335.
Refs #50330 (TODO for better error messaging)

Release note (sql change): Add a cluster setting
`sql.conn.max_read_buffer_message_size` that allows users to configure
sql statement maximum sizes across the cluster. This effect will take
place upon connection restart.


Release justification: low risk, high benefit changes to existing
functionality

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53445)
<!-- Reviewable:end -->

otan committed:
sql: add cluster setting for maximum read buffer size

Release note (sql change): Add an invisible cluster setting
`sql.conn.max_read_buffer_message_size` that allows users to configure
sql statement maximum sizes across the cluster. This effect will take
place upon connection restart.

Release justification: low risk, high benefit changes to existing
functionality

solongordon commented:
Oh, this needs a "Release justification" in the commit message.

otan commented:
in the commit message but not PR description, heh

otan commented:
done (it was in commit msg but not PR description)

bors r=solongordon

otan commented:
bors r=solongordon

otan commented:
bors r=solongordon

otan commented:
bors r=solongordon
